### DISCUSSION TITLE
Restructuring a GLTF/GLB - Issues with Coping Some Meshes (Most Work)

### DISCUSSION THREAD

**Discussion by Jonathynlee** [donmccurdy/glTF-Transform#1328] [General]
Hello Don! 
Hope you are well. 

I am working on an app that imposes a new structure on the gltf. We call this the "Sales GLTF" and the original the "3D Asset Gltf" (it comes from CAD). 

These are the steps I am taking:
1) Traverse the original document (3d Asset Gltf), create a map  {sUUID : Node}  //This sUUID is a business uuid the we define, not the gltf uuid.  
2) Traverse our artificial BOM and copy the nodes in a recursive fashion using the map above as the source. 
```typescript
const resolve = (source) => {   
        if (source instanceof Node) {   
          return salesBomDocument.createNode().copy(source, resolve);   
        }   
        return source.clone();   
};   
const traverseBom:(orginalMap, currentSalesNode, currentDocumentNode, currentSalesNodeAssociatedUUID)=>{   
   
   const newNode = document.createNode().copy(originalMap[currentSalesNodeAssociatedUUID], resolve);   
   currentSalesNode.children.forEach((aChildSalesNode)=>{
      traverseBom:(orginalMap, aChildSalesNode, newNode, currentSalesNodeAssociatedUUID)
   }
   currentDocumentNode.addChild(newNode);   
}   
```

3) Remove all traces of original nodes
```typescript
for(let aNodeSUUID in orginalMap){   
  mainScene.removeChild(orginalMap[aNodeSUUID ])   
  orginalMap[aNodeSUUID ].dispose();   
  delete orginalMap[aNodeSUUID ];   
}   
```

Most of everything seems fine, but it seem that a few of the new nodes are not coping the Mesh from the original node. All of the metadata (name, extras, rot, scale, trans) is being transferred properly. 

Here is one of the nodes that is not copying the mesh (just console.log'd)
![image](https://github.com/donmccurdy/glTF-Transform/assets/54726329/9762d791-bedf-4fef-96ae-1279bb571d09)

How it looks in the original model
![image](https://github.com/donmccurdy/glTF-Transform/assets/54726329/055e234f-d128-4b63-a5bf-a6a0905b8182)

How it looks in the newly structured model
![image](https://github.com/donmccurdy/glTF-Transform/assets/54726329/c5a6f014-cb21-40c9-8cd7-d21e8d54c47c)

There is also some duplication of mesh as well but I havn't tackled that problem yet. I have been scratching my head over this for hours and figured it was time to ask the legend!

Thanks in advanced!
-Jonathyn

**Comment by Jonathynlee** [comment]
Also @donmccurdy , 
I am happy to share more details about the code and project if needed in a private chain. This is for a client so I can't put everything here.

**Comment by Jonathynlee** [comment]
After reviewing further, it looks like I was not more careful in my traverse and I was pulling over subnodes of assemblies, For the missing pieces, it looks like that might be an issue with our viewing tech. Still not sure what changed, but I pulled this model into gltf-report and the bolts do show up. 

We know your time is valuable so sorry for the false alarm and thanks!
-Jonathyn

**Reply by donmccurdy** [reply]
Hi @Jonathynlee! One other thought that might be helpful if similar issues come up again. The library doesn't currently offer an easy way to transfer or copy objects from one Document to another (see #924 for that), but you're on the right track with implementing a `resolve` callback here.

Probably that `resolve` function should never clone an object from the source document, but should always (1) check for a cached copy in a Map, (2) create a new object and copy the source object's properties if no cached copy is found. That's essentially how `document.clone()` is implemented, which might be helpful as a reference:

https://github.com/donmccurdy/glTF-Transform/blob/c47cbd2ed048d94ff35ee46a46e814fb89088a38/packages/core/src/document.ts#L136-L142
